### PR TITLE
Change base_ring -> coefficient_ring for polys

### PR DESCRIPTION
**Pull request by thofma**
Includes (M/Sparse/NC)Poly, Laurent, Puiseux.

Should I also do Series?

### CODE REVIEW

File: src/MPoly.jl [hunk 1]
```diff
@@ -808,7 +808,7 @@ function _evaluate(a, S, R, vars, vals::Vector{Rational{BigInt}})
 end
 
 function __evaluate(a, vars, vals, powers)
-   R = base_ring(a)
+   R = parent(a) isa MSeriesRing ? base_ring(a) : coefficient_ring(a)
```
Comment (thofma) [inline src/MPoly.jl:811]: Apparently some `MSeriesRing` code calls `__evaluate`. Since I did not change for series, it now looks like this.

**Comment by wbhart** [comment]
I'm not sure either way. I thought the main thing we wanted it for was multivariate polynomials.

Do you really do this for PuiseuxSeries and LaurentSeries and not AbsSeries and RelSeries?

What will be the distinction between base_ring and coefficient_ring in the general case? I.e. what definition are you using?

**Comment by wbhart** [comment]
Feel free to merge when you are satisfied.

**Comment by thofma** [comment]
I started with multivariate poly, but then realised that it would be funny to have it for multivariate but not univariate polynomial rings. I will roll it back for Laurent series and Puiseux series.

The definition? I guess the R in R[x1,...,xn] is the coefficient ring. Similar definition would apply to any series ring. I think it is a well defined concept.

Alternatively we could have both, coefficient and base ring, for polynomial rings.

**Comment by wbhart** [comment]
It seems strange to not have base_ring = coefficient_ring for generic polynomials. But I leave it to your judgement how to handle it.

**Comment by thofma** [comment]
Can you say again what your reasons were for suggesting coefficient_ring?

**Comment by wbhart** [comment]
My example was rational function fields k(x) represented as a fraction field of k[x]. It makes sense (I think) for k[x] to be the base_ring and k to be the coefficient field.

As things are, base_ring is defined programmatically for generic things defined as a mathematical construction over another ring. But as you say, it is sometimes convenient to have a coefficient ring regardless of the way the object was constructed.

An example is Singular which has ideals over polynomial rings over a field. The coefficient_ring would be the field, but the base ring of the ideals would be the polynomial ring.

**Comment by wbhart** [comment]
I think the main thing we wanted to solve was that base_ring of an fmpz_mpoly is Union{} for example. This breaks any generic code that really wants to deal with the coefficient ring.

**Comment by thofma** [comment]
Does your last comment mean that we *want* to have `base_ring(::fmpz_mpoly) = Union{}`? Because at the moment it is not.

**Comment by wbhart** [comment]
Oh sorry, I thought it was. To be honest, I am not sure what we want.

**Comment by thofma** [comment]
There is a wish (https://github.com/oscar-system/Oscar.jl/issues/454) to have `coefficient_ring` in addition to `base_ring` (where it makes sense).

So lets scrap this pull request here and discuss what we want.

I think we definitely want to add the `coefficient_ring` function here in AA, even if we just add the name. Then going from there, there are two options. Either we add the `coefficient_ring(...) = base_ring(...)` alias here or we will do it somewhere else. (But it will definitely be added.)

P.S.: Regarding nomenclature for function fields, see http://magma.maths.usyd.edu.au/magma/handbook/text/458#4746.

**Comment by wbhart** [comment]
Yeah, I think that what we probably want for now is coefficient_ring(a::PolyElem) = base_ring(a) and coefficient_ring(a::MPolyElem) = base_ring(a).

That's probably sufficient.

Number fields are perhaps problematic in that base_ring(a::AnticNumberField) = Union{} (I hope I have it right this time). Probably we should make coefficient_ring(a::AnticNumberField) == QQ

It looks like Magma defines coefficient_ring for a rational function field the way I want to define it. So that is at least good. I was unaware that the polynomial ring was called the ring of integers. I'll ask @fieker if he is happy with that name and I can adjust it in the existing code if so.

**Pull request closed without merging**